### PR TITLE
OpenAPI/swagger docs autogen

### PR DESCRIPTION
**Pull request by yaleman**
Refers to #2028, starting to implement an OpenAPI schema generation for the server. Currently testing with `utoipa` but wow there's a lot of boilerplate and macros and stuff.

Checklist

- [x] This pr contains no AI generated code
- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes

### CODE REVIEW

**Comment by yaleman** [comment]
But hey, it happens quickly... 😄 

Screenshot 2023-10-04 at 21 39 37

File: server/core/src/https/v1.rs [hunk 1]
```diff
@@ -38,7 +37,22 @@ pub(crate) struct SessionId {
     pub sessionid: Uuid,
 }
 
-// /v1/raw/create
+#[utoipa::path(
+    post,
+    path = "/v1/raw/create",
```
Comment (Firstyear) [inline server/core/src/https/v1.rs:42]: I like that you added this with openapi, but lets *not* document the /raw commands ... they are there as breakglass for us to issue emergency help commands if needed.

File: server/core/src/https/v1_scim.rs [hunk 2]
```diff
@@ -151,6 +183,23 @@ pub async fn sync_account_id_put_attr(
     json_rest_event_put_id_attr(state, id, attr, filter, values, kopid).await
 }
 
+#[utoipa::path(
+    get,
+    path = "/scim/v1/Sink",
```
Comment (Firstyear) [inline server/core/src/https/v1_scim.rs:188]: This is an easter egg, should we document it?

**Review by Firstyear — CHANGES_REQUESTED** [review]
I think my biggest concert here is this looks pretty noisy/manual to implement. Should we limit this to apis we really want people to interact with?

File: server/core/src/https/v1.rs (hunk 1 above)
Comment (yaleman) [reply server/core/src/https/v1.rs:42]: Done

**Comment by yaleman** [comment]
> I think my biggest concert here is this looks pretty noisy/manual to implement. Should we limit this to apis we really want people to interact with?

The biggest problem I've had trying to implement and work with the Kanidm APIs is the lack of documentation and typing (no judgement, just reality) on input/output, so ... IMO a clearly documented API is needed for **_everything_** to make the DX a lot smoother. 

I agree it's pretty bulky, and I think there's some things that can make it neater with the `axum` integration, but it's much less bad than manually building an OpenAPI schema 🤣

**Comment by yaleman** [comment]
This is as far as I'm going to get at this point, lots of TODOs but WAY better docs for API endpoints etc. and less use of unwrap and expect

**Comment by Firstyear** [comment]
Ready for review?

**Comment by yaleman** [comment]
> Ready for review?

Sure is!

File: server/core/src/https/errors.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+/// The web app's top level error type, this takes an `OperationError` and converts it into a HTTP response.
+#[derive(Debug, ToSchema)]
+pub enum WebError {
+    /// Something went wrong when doing things.
+    OperationError(OperationError),
+    InternalServerError(String),
+}
+
+impl From<OperationError> for WebError {
```
Comment (Firstyear) [inline server/core/src/https/errors.rs:18]: Is this kind of tied into the error handling stuff we want to do in future?

**Review by Firstyear — APPROVED** [review]

File: server/core/src/https/errors.rs (hunk 3 above)
Comment (yaleman) [reply server/core/src/https/errors.rs:18]: Yep, it's a look at how that kind of thing could work in the future.

**Pull request merged**